File: sort/InsertSort.js
#!/bin/node


/**
 * Insertion Sort.
 *
 * zhaoyou@gmail.com (zhaoyou)
 */

function main() {
  var datas = [3, 7, 4, 2, 8,  9 ];
  console.log(InsertSort(datas));
}

function swap(data, from , to) {
  var temp = data[from];
  data[from] = data[to];
  data[to] = temp;
}

function InsertSort(datas) {
  var i, j, temp;
  var len = datas.length;
  for (i = 1; i < len; i++) {
    if ( datas[i] < datas[i-1]) {
      temp = datas[i];
      j = i;

      while (j > 0 && datas[j - 1] > temp) {
        datas[j] = datas[j-1];
        j--;
      }
      datas[j] = temp;
    }
  }

  return datas;
}

main();

File: Sort/InsertSort.java
public class InsertSort {

	public static void main(String[] args) {
		int array[] = new int[]{3, 7 , 8, 1, 4, 9};

		for(int i = 1; i < array.length; i++) {
			for (int k = i; k > 0 && array[k] < array[k - 1]; k--) {
				swap(array, k, k -1 );
			}
		}

		for(int j = 0; j < array.length; j++) {
			System.out.println(array[j]);
		}
	}

	public static void swap(int[] array, int first, int last) {
		int temp = array[first];
		array[first] = array[last];
		array[last] = temp;
	}
}